SimpleTerminologyLoadTest.ipynb

```diff
@@ -0,0 +1,331 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "266cc6d5",
+   "metadata": {},
+   "source": [
+    "# Simple Terminology Load Test\n",
+    "this is wall-clock time where a server-side cache may help and give better results\n",
```

Comment:
Noted

Reply:
so, wall-clock is subject to all kinds of bias or interference like other things running on the server and network latency. I haven't looked at what Python has for proper "profiling", but in C and Java and very likely Python too,  there's ways of counting instructions executed no matter the server load as a more direct way of looking at the algorithms in the code. Which is to say wall-clock measures like this are amateurish and looked-down upon. In this case, I'm claiming some practical value that includes the server load on Azure and network latency, etc.

Reply:
I don't have a lot of experience in performance profiling, unfortunately. I think this is a fine way to do it. In Python, I just always use the pattern:

```py
from datetime import datetime
t1 = datetime.now()
# code I'm timing goes here
t2 = datetime.now()
print(f'Finished in {(t2 - t1).seconds} seconds')
```

Reply:
a person could play around with repeating that pattern and looking to see how consistent it is. You wouldn't want to compare two things and have the bad luck of a system process kick in right when you're running the second. 

Proper profiling gets closer to "big oh" characterizations of algorithms. I don't know how much of this they teach in CS anymore. We had to write our data structures in C and then tell why quick sort or merge sort is faster than bubble sort: O(n log(n))  vs O(n^2) . Basically counting how many instructions an algorithm has.  

But professionally? I think I've actually pulled out a profiler two or three times in 30+ years.

Reply:
(fun): I wrote bubble and merge sorts in python and compared their performance to the library function on 10k integers. Outputs are wall-clock time in microseconds and the estimate of loop iterations: n^2 for bubble sort and n-log(n) for merge sort. The way I wrote it, one loop in merge sort takes longer than bubble (6x), but so fewer steps it comes out ahead. The motivation for me was to see how much the compiled library sort beats interpreted code: ~30x.
Library:: 1.177
Bubble: 6056.738, 100000000
Merge: 39.239000000000004, 132877.1237954945
Bubble:Merge  time ratio: 154.35505491985015, steps ratio: 752.574989159953
Merge:Library  time ratio: **33.3**3814783347494

Reply:
(more fun): the profiler breaks out the functions, and doesn't give as clear insight to the number of steps as I might of thought. I might guess that len(array) counts the length instead of there being an attribute...


  ```
       738079 function calls (718081 primitive calls) in 0.991 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   133616    0.086    0.000    0.086    0.000 :0(append)
        1    0.000    0.000    0.991    0.991 :0(exec)
   574461    0.364    0.000    0.364    0.000 :0(len)
        1    0.000    0.000    0.000    0.000 :0(setprofile)
        1    0.000    0.000    0.991    0.991 <string>:1(<module>)
        1    0.000    0.000    0.991    0.991 profile:0(d=merge_sort(d))
        0    0.000             0.000          profile:0(profiler)
     9999    0.478    0.000    0.911    0.000 sort.py:46(merge)
  19999/1    0.063    0.000    0.990    0.990 sort.py:65(merge_sort)
```

using a variable for len(array), got me a bit of improvement to
Merge: 26.947, 132877.1237954945
```          203615 function calls (183617 primitive calls) in 0.328 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   133616    0.089    0.000    0.089    0.000 :0(append)
        1    0.000    0.000    0.328    0.328 :0(exec)
    39997    0.033    0.000    0.033    0.000 :0(len)
        1    0.000    0.000    0.000    0.000 :0(setprofile)
        1    0.000    0.000    0.328    0.328 <string>:1(<module>)
        1    0.000    0.000    0.328    0.328 profile:0(d=merge_sort(d))
        0    0.000             0.000          profile:0(profiler)
     9999    0.144    0.000    0.250    0.000 sort.py:46(merge)
  19999/1    0.062    0.000    0.328    0.328 sort.py:70(merge_sort)